### PR TITLE
Added Concat Operation

### PR DESCRIPTION
**Pull request by cs-tushar-kanade**
Added Concat Operation
Updated tool tip for inputs

### CODE REVIEW

File: csv-data-management/extract_data_from_csv.py [hunk 1]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
         logger.error('Error in extract_data_from_csv(): %s' % Err)
 
 
+def concat_two_csv_and_extract_data(config, params):
+    try:
+        numberOfRowsToSkipFirst = None
+        numberOfRowsToSkipSecond = None
+        isSingleColumnFrist = None
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: Frist --> First

File: csv-data-management/extract_data_from_csv.py [hunk 2]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        isSingleColumnSecond = None
+        isCSVWithoutHeadersFirst = False
+        isCSVWithoutHeadersSecond = False
+
+        file_1_iri_ = handle_params(params,params.get('file_one_value'))
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: file_1_iri --> fileOneIri
file_2_iri --> fileTwoIri

File: csv-data-management/extract_data_from_csv.py [hunk 3]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        file_2_iri_ = handle_params(params,params.get('file_two_value'))
+        logger.warning(params)
+        first_file_path = join('/tmp', download_file_from_cyops(file_1_iri_)['cyops_file_path'])
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: fileFirstPath
fileTwoPath

File: csv-data-management/extract_data_from_csv.py [hunk 4]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        second_file_path = join('/tmp', download_file_from_cyops(file_2_iri_)['cyops_file_path'])
+
+        res = _check_if_csv(first_file_path)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: _check_if_csv indicates that it returns bool? if yes, then we should check for T/F as well and bailout

File: csv-data-management/extract_data_from_csv.py [hunk 5]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        logger.info(res)
+  
+        if res.get('headers') == False:
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: Is it a common code among multiple funs? if yes, we can create a Init kind func to init all values/flags

File: csv-data-management/extract_data_from_csv.py [hunk 6]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        
+        res = _check_if_csv(second_file_path)
+
+        if res.get('headers') == False:
+            isCSVWithoutHeadersSecond = True
+        elif res.get('columns') == 1:
+            isSingleColumnSecond = True
+
+
+        #Lets read first file
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: Can be converted into a small func and called for both files one by one

File: csv-data-management/extract_data_from_csv.py [hunk 7]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        elif res.get('columns') == 1:
+            isSingleColumnSecond = True
+
+
+        #Lets read first file 
+         
+        if params.get('numberOfRowsToSkipFirst'):
+          numberOfRowsToSkipFirst = params.get('numberOfRowsToSkipFirst')
+          
+        if params.get('file1_columnNames') != "":  # CSV file with column header and specifci columns to use in creating recordset
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: specifci --> specific

Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: file1_columnNames --> file1_column_names

Comment (cs-suyog-jinturkar) [reply csv-data-management/extract_data_from_csv.py]: file_one_column_names OR file_one_col_names

File: csv-data-management/extract_data_from_csv.py [hunk 8]
```diff
@@ -200,6 +200,119 @@ def merge_two_csv_and_extract_data(config, params):
...
+        combined_recordSet =pd.concat([df_file1,df_file2])    
+
+        # If user has selected to dedeuplicate recordset
+        try:
+            if params.get('deDupValuesOn'):
+                deDupValuesOn = params.get('deDupValuesOn')
+                deDupValuesOn = deDupValuesOn.split(",")
+                combined_recordSet.drop_duplicates(subset=deDupValuesOn, keep='first')
+        except Exception as Err:
+            logger.error('Error in dedeuplicating data  extract_data_from_csv(): %s' % Err)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: function name will be present in stack trace.

Comment (cs-suyog-jinturkar) [reply csv-data-management/extract_data_from_csv.py]: but, why we are still continuing ahead with exception?

File: csv-data-management/info.json [hunk 9]
```diff
@@ -44,7 +44,8 @@
           "visible": true,
           "editable": true,
           "value": "",
-          "description": "comma separate lis of columns name you want to extract from CSV File"
+          "description": "comma separate lis of columns name you want to extract from CSV File",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: lis --> list

Comment (cs-suyog-jinturkar) [reply csv-data-management/info.json]: comma separated column names you want to extract from a CSV file

File: csv-data-management/info.json [hunk 10]
```diff
@@ -54,7 +55,8 @@
           "visible": true,
           "editable": true,
           "value": "",
-          "description": "Column name you want to use deduplicate date  from CSV File"
+          "description": "Column name you want to use deduplicate date  from CSV File",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json:58]: Didn't understand this desc

Comment (cs-suyog-jinturkar) [reply csv-data-management/info.json:58]: how about, Use this column name for deduplication?

File: csv-data-management/info.json [hunk 11]
```diff
@@ -74,7 +77,8 @@
           "visible": true,
           "editable": true,
           "value": null,
-          "description": "If set , will return rows as recordSet in list of 20 batches otherwise complete result in a single variable"
+          "description": "If set , will return rows as recordSet in list of 20 batches otherwise complete result in a single variable",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json:80]: which variable?

File: csv-data-management/info.json [hunk 12]
```diff
@@ -197,7 +208,126 @@
...
+          "name": "file_two_value",
+          "required": true,
+          "visible": true,
+          "editable": true,
+          "value": ""
+        },
+        {
+          "title": "Second File Column Names",
+          "type": "text",
+          "name": "file2_columnNames",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: Please use consistent names

Comment (cs-suyog-jinturkar) [reply csv-data-management/info.json]: camel case of with underscore

File: csv-data-management/extract_data_from_csv.py (hunk 2 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: I will stay with current nomeclature if is not required , from readability perspective

File: csv-data-management/extract_data_from_csv.py (hunk 3 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: I will stay with current nomenclature if is not required , from readability perspective

File: csv-data-management/extract_data_from_csv.py (hunk 4 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: no its not boolean , its return {"headers": True,"columns": col }

File: csv-data-management/extract_data_from_csv.py (hunk 5 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: Will take it up in next iteration , will require some code layout change

File: csv-data-management/extract_data_from_csv.py (hunk 6 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: Will take it up in next iteration , will require some code layout change

File: csv-data-management/extract_data_from_csv.py (hunk 7 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: used - file1_column_names

File: csv-data-management/extract_data_from_csv.py (hunk 8 above)
Comment (cs-tushar-kanade) [reply csv-data-management/extract_data_from_csv.py]: Statements in logs are more clearer , and we can search better as single connector log is used to write for all connectors 

With exception thrown code terminates , did not see it continuing , but will re-test

File: csv-data-management/info.json (hunk 12 above)
Comment (cs-tushar-kanade) [reply csv-data-management/info.json]: done

File: csv-data-management/builtins.py [hunk 13]
```diff
@@ -1,4 +1,4 @@
 #FSR Autogenerated Content. DO NOT DELETE
 from .extract_data_from_csv import *
  
-supported_operations = {'extract_data_from_csv': extract_data_from_csv, 'merge_two_csv_and_extract_data': merge_two_csv_and_extract_data }
\ No newline at end of file
+supported_operations = {'extract_data_from_csv': extract_data_from_csv, 'merge_two_csv_and_extract_data': merge_two_csv_and_extract_data, 'concat_two_csv_and_extract_data': concat_two_csv_and_extract_data }
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/builtins.py:4]: Can't we just call these operations as "merge_csv_files", "concat_csv_files"?

File: csv-data-management/extract_data_from_csv.py [hunk 14]
```diff
@@ -87,90 +87,63 @@ def merge_two_csv_and_extract_data(config, params):
...
-        elif res.get('columns') == 1:
-            isSingleColumnSecond = True
-
+        fileOneIRI = handle_params(params,params.get('file_one_value'))
+        fileOnePath = join('/tmp', download_file_from_cyops(fileOneIRI)['cyops_file_path'])
+        fileTwoIRI = handle_params(params,params.get('file_two_value'))
+        fileTwoPath = join('/tmp', download_file_from_cyops(fileTwoIRI)['cyops_file_path'])
+        logger.info(params)
+        # Read First File
+        df1 = _read_and_return_ds(fileOnePath,params,config)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:96]: return_ds or return_df?

Comment (cs-suyog-jinturkar) [reply csv-data-management/extract_data_from_csv.py:96]: Generally, we don't publish return type in func name.

File: csv-data-management/extract_data_from_csv.py [hunk 15]
```diff
@@ -87,90 +87,63 @@ def merge_two_csv_and_extract_data(config, params):
...
-        elif isSingleColumnFrist and not isCSVWithoutHeadersFirst: # CSV file with one  column and no header 
-            df_file1 = _read_file_single_column_no_header(first_file_path,numberOfRowsToSkipFirst)
+        # If user has selected to dedeuplicate recordset
+        try:
+            if params.get('deDupValuesOn'):
+                deDupValuesOn = params.get('deDupValuesOn')
+                deDupValuesOn = deDupValuesOn.split(",")
+                combined_recordSet.drop_duplicates(subset=deDupValuesOn, keep='first')
+        except Exception as Err:
+            logger.error('Error in dedeuplicating data  extract_data_from_csv(): %s' % Err)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: dedeuplicating --> Deduplicating

File: csv-data-management/extract_data_from_csv.py [hunk 16]
```diff
@@ -87,90 +87,63 @@ def merge_two_csv_and_extract_data(config, params):
...
-            df_file1 =  _read_file_no_headers(first_file_path,numberOfRowsToSkipFirst) 
+        #Merge both files
+        combined_recordSet =pd.merge(df1,df2,how='left',left_on=mergeColumn,right_on=mergeColumn)    
 
-        elif isSingleColumnFrist: #CSV with single column and header    
-            df_file1 = _read_file_single_column(first_file_path,numberOfRowsToSkipFirst)
-        
-        elif isSingleColumnFrist and not isCSVWithoutHeadersFirst: # CSV file with one  column and no header 
-            df_file1 = _read_file_single_column_no_header(first_file_path,numberOfRowsToSkipFirst)
+        # If user has selected to dedeuplicate recordset
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: dedeuplicate --> Deduplicate

File: csv-data-management/extract_data_from_csv.py [hunk 17]
```diff
@@ -87,90 +87,63 @@ def merge_two_csv_and_extract_data(config, params):
...
+        try:
+            if params.get('deDupValuesOn'):
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:105]: why .get()? can't we use .contains() or so here?

File: csv-data-management/extract_data_from_csv.py (hunk 15 above)
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py]: dedeuplicating --> Deduplicating

File: csv-data-management/extract_data_from_csv.py [hunk 18]
```diff
@@ -87,90 +87,63 @@ def merge_two_csv_and_extract_data(config, params):
...
-        if params.get('numberOfRowsToSkipSecond'):
-            numberOfRowsToSkipSecond = params.get('numberOfRowsToSkipSecond')
-          
-        if params.get('file2_columnNames') != "":  # CSV file with column header and specifci columns to use in creating recordset 
-            columnNames = params.get('file2_columnNames')
-            columnNames = columnNames.split(",")
-    
-            # We are passing  specific columns name to filter data from here
-            df_file2 = _read_file_specific_columns(second_file_path,columnNames,numberOfRowsToSkipSecond)
+        #Create small chunks of dataset to cosume by playbook if requested by user otherwise return complete recordset
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:115]: cosume --> consume

File: csv-data-management/extract_data_from_csv.py [hunk 19]
```diff
@@ -237,22 +209,29 @@ def _read_file_single_column(filepath,numberOfRowsToSkip=None):
...
         logger.error('Error in _read_file_no_headers(): %s' % Err) 
             
-def _read_file_single_column_no_header(filepath,numberOfRowsToSkip=None):
+def _read_file_single_column_no_header(filepath,numberOfRowsToSkip=None,no_of_columns=None):
     try:
+        if no_of_columns:
+            colList = []
+            for i in range(no_of_columns):
+                colList.append("Column"+str(i))
         chunk = pd.read_csv('{}'.format(filepath),usecols=[0],header = None,skiprows=numberOfRowsToSkip,chunksize=100000,error_bad_lines=False)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:218]: chunksize --> can it init at global level and used here as a config param?

File: csv-data-management/extract_data_from_csv.py [hunk 20]
```diff
@@ -237,22 +209,29 @@ def _read_file_single_column(filepath,numberOfRowsToSkip=None):
...
 def _check_if_csv(filepath):
     try:
         sniffer = csv.Sniffer()
-        sample_bytes = 32
-        res=sniffer.has_header(open(filepath).read(sample_bytes))
+        res = sniffer.has_header(open(filepath).read(2048))
         df = pd.read_csv('{}'.format(filepath),error_bad_lines=False,nrows=10)
         row, col = df.shape
-        return {"headers": True,"columns": col }
+        if  res:
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:231]: This if block should be just after res = sniffer.has, right?

File: csv-data-management/extract_data_from_csv.py [hunk 21]
```diff
@@ -237,22 +209,29 @@ def _read_file_single_column(filepath,numberOfRowsToSkip=None):
...
     except Exception as Err:
         logger.error('Error in _read_file_single_column_no_header(): %s' % Err) 
 
 def _check_if_csv(filepath):
     try:
         sniffer = csv.Sniffer()
-        sample_bytes = 32
-        res=sniffer.has_header(open(filepath).read(sample_bytes))
+        res = sniffer.has_header(open(filepath).read(2048))
         df = pd.read_csv('{}'.format(filepath),error_bad_lines=False,nrows=10)
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/extract_data_from_csv.py:229]: What is pd? what is df?

Comment (cs-suyog-jinturkar) [reply csv-data-management/extract_data_from_csv.py:229]: what is ds?

File: csv-data-management/info.json [hunk 22]
```diff
@@ -109,12 +113,13 @@
...
           "title": "First File Column Names",
           "type": "text",
-          "name": "file1_columnNames",
+          "name": "file1_column_names",
           "required": false,
           "visible": true,
           "editable": true,
           "value": "",
-          "description": "comma separate lis of columns name you want to extract from CSV File"
+          "description": "comma separate lis of columns name you want to extract from CSV File",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: lis --> list

File: csv-data-management/info.json [hunk 23]
```diff
@@ -151,12 +158,13 @@
...
           "title": "Second File Column Names",
           "type": "text",
-          "name": "file2_columnNames",
+          "name": "file2_column_names",
           "required": false,
           "visible": true,
           "editable": true,
           "value": "",
-          "description": "comma separate lis of columns name you want to extract from CSV File"
+          "description": "comma separate lis of columns name you want to extract from CSV File",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: lis --> list

File: csv-data-management/info.json [hunk 24]
```diff
@@ -151,12 +158,13 @@
...
+          "tooltip":"comma separate lis of columns name you want to extract from CSV File"
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: lis --> list

File: csv-data-management/info.json [hunk 25]
```diff
@@ -197,7 +208,126 @@
           "visible": true,
           "editable": true,
           "value": null,
-          "description": "If set , will return rows as recordSet in list of 20 batches otherwise complete result in a single variable"
+          "description": "If set , will return rows as recordSet in list of 20 batches otherwise complete result in a single variable",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json:211]: Which variable?

File: csv-data-management/info.json [hunk 26]
```diff
@@ -197,7 +208,126 @@
...
+          "tooltip":"If set , will return rows as recordSet in list of 20 batches otherwise complete result in a single variable"
+        }
+      ],
+      "open": true
+    },
+    {
+      "title": "Concat and Extract Data from two CSV",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json:218]: CSV --> CSVs

File: csv-data-management/info.json [hunk 27]
```diff
@@ -197,7 +208,126 @@
...
+      "operation": "concat_two_csv_and_extract_data",
+      "description": "This action will extract data by concatting two CSV file. User can select option to deduplicate rows based on columns",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json:220]: file --> files

File: csv-data-management/info.json [hunk 28]
```diff
@@ -197,7 +208,126 @@
...
+        },
+        {
+          "title": "First File Column Names",
+          "type": "text",
+          "name": "file1_column_names",
+          "required": false,
+          "visible": true,
+          "editable": true,
+          "value": "",
+          "description": "comma separate lis of columns name you want to extract from CSV File",
```
Comment (cs-suyog-jinturkar) [inline csv-data-management/info.json]: lis --> list

**Review by cs-suyog-jinturkar — COMMENTED** [review]
Added review comments

**Review by cs-suyog-jinturkar — APPROVED** [review]

**Pull request merged**